README.md
# KVNBoundedImageView
KVNBoundedImageView attempts to keep faces visible and centered in a UIImageView. It is designed to be easy to use and extensible for different types of image detection.

#### From
![Disabled ImageView](https://raw.github.com/donnellyk/KVNBoundedImageView/master/Assets/disabled.png) 
#### To 
![Enabled](https://raw.github.com/donnellyk/KVNBoundedImageView/master/Assets/enabled.png) 

With configurable detection speed, operation queues, caching, and utilizing the optimized image rendering of UIImageView (no custom drawing here), KVNBoundedImageView aims to be as fast as possible, without blocking the main thread when heavy lifting is needed. Simple image loading via a URL is also available, so you may even be able to throw away that UIImageView category you've been dragging along between projects.

[![Version](http://cocoapod-badges.herokuapp.com/v/KVNBoundedImageView/badge.png)](http://cocoadocs.org/docsets/KVNBoundedImageView)
[![Platform](http://cocoapod-badges.herokuapp.com/p/KVNBoundedImageView/badge.png)](http://cocoadocs.org/docsets/KVNBoundedImageView)

## Usage
KVNBoundedImageView attempts to watch as many changes as possible within the UIImageView's properties and respond appropriately to keep the features visible. However, if at any point you find the view not responding to changes, you can manually force it to update by calling `fitToFeature`. See the Example project for more details on all the examples below


### Creating from NIB or Storyboard
Drag a UIImageView onto the scene and change it's class to KVNBoundedImageView in the Utilities pane under Identity Inspector. You can set the reuse cache name under User Defined Runtime Attributes by setting `nibImageCacheName` with a NSString value.

![Utility Pane screenshot](https://raw.github.com/donnellyk/KVNBoundedImageView/master/Assets/utility.png)


### Creating Programically
#### Bare Minimum
```objective-c
KVNBoundedImageView *imageView = [[KVNBoundedImageView alloc] initWithFrame:aRect];
[imageView setImage:[UIImage imageNamed:@"test"] cacheName:@"test"];
    
[imageScrollView addSubview:imageView];
```


#### With Some Configuration

When doing custom configuration, it is advised to set the all the parameters before you set the image to be detected. Otherwise, the detection & cropping will start, and then immediately be cancelled and restarted.   

```objective-c
KVNBoundedImageView *imageView = [[KVNBoundedImageView alloc] initWithFrame:aRect];
[imageView setBoundingBoxScheme:BoundingBoxSchemeLargest];
[imageView setBoundingPadding:10.0];
[imageView setImage:[UIImage imageNamed:@"test"] cacheName:@"test" placeholder:[UIImage imageNamed:@"placeholder"]];
    
[imageScrollView addSubview:imageView];
```


#### Load from a URL

```objective-c
KVNBoundedImageView *imageView = [[KVNBoundedImageView alloc] initWithFrame:aRect];    
[imageScrollView setImageFromURL:aURL cacheName:@"test" placeholder:[UIImage imageNamed:@"placeholder"]];
```


## Properties

- `boundingPadding`: `CIDetector` has a thing against foreheads, it doesn't include them in the detection bounds. If you roughly know the size of the faces in your image or if you have having trouble with foreheads being cut off at the top of your image (or chins at the bottom), playing with this value might give you better results. It isn't a perfect science, though, and edge-cases abound!
- `detectorAccuracy`: Changes the `CIDetectorAccuracy` used for feature detection. Default is `CIDetectorAccuracyLow`
- `boundingBoxScheme`: How the bounding rectangle is calculated.
    - `BoundingBoxSchemeAll`: All features detected are used. A rectangle that fits all features is created and used. This is the default behavior
    - `BoundingBoxSchemeLargest`: Uses the largest face found. Good for if you have images of crowds with a clear subject.
    - `BoundingBoxSchemeSmallest': Uses the smallest face found. To be honest, I only threw this in because it was super easy to implement. I have no idea why you would want to use this. Maybe because everyone loves an underdog?
- `boundingEnabled`: Enabled or disables the detection and bounding. If `NO`, the image is displayed with `UIViewContentModeScaleAspectFill`
- `animated`: To make the transition a little nicer (but only a little, there is room for improvement), this provides a short Core Animation (kCATransitionFade) when the detection has finished. Defaults to `YES`.

## Installation

To manually install KVNBoundedImageView, add KVNBoundedImageView.m and KVNBoundedImageView.h files to your project. Import the CoreImage framework into your project.

### Installation with CocoaPods 

[CocoaPods](http://cocoapods.org) is a dependency manager for Objective-C, which automates and simplifies the process of using 3rd-party libraries like KVNBoundedImageView in your projects. See the ["Getting Started" guide for more information](https://github.com/AFNetworking/AFNetworking/wiki/Getting-Started-with-AFNetworking).

#### Podfile

```ruby
pod "KVNBoundedImageView"
```

### Requirements
- CoreImage
- iOS 5.0+

## Extending KVNBoundingImageView

This library is designed with extensibility in mind. It is fairly easy to implement your own detection, so long as you do the actual detection leg work yourself (IE: The Hard Part). Perhaps you have written a dog detector using `OpenCV` and want to make an app focused on dogs, without pesky cats getting in the way (A noble calling). To accomplish this, all you have to do is implement `boundingRectangleForImage` in your subclass and return the `CGRect` that you wish to be visible. This rect needs to be in the `UIView` coordinate system (origin is in the upper-left). Keep in mind that `boundingRectangleForImage` may be called on a background queue, so operations that are not thread safe should be done. As a convenience to developers who may be using your subclass, you should also extend `foundFeatures` and return all the rects (stored in `NSValue`) that were detected in the image.

## Note on Optimization and Caching

This library utilizes a NSCache to store the cropped images for fast recall. The key uses a supplied cache name along with the detection accuracy, bounding box scheme, and current view aspect ratio. This is so the image does not get reused in a view that is won't properly fit.  NSCache responds to OS-level memory warnings to clear space, but if you are worried about memory usage, minimize the different configuration you use or you can just use `setImage:` or pass nil for a cache name and nothing will be cached. This caching designed with use in a UITableView in mind, however it hasn't been extensively tested so proceed with caution.

### Memory Usage

This library uses a bit more memory then one would think, mainly around the usage of CIDetector. CIDetector seems to have a problem with unbounded memory growth. Not reusing the detector, instead instantiating one as needed, mitigates this problem but still doesn't seem to remove it entirely and is against what the documentation recommends. [OpenRadar](http://openradar.appspot.com/16061776) concerning this issue.

### Animations

The imageview has to recrop the image with every aspect ratio change, so it probably not the best idea to have this component enabled while animating the bounds (center, origin is fine). The recommended way:

```objective-c
UIImage *originalImage = imageView.originalImage;
UIImage *croppedImage = imageView.image;

[imageView setBoundingEnabled:YES];
[imageView setImage:croppedImage];
[UIView animateWithDuration:1.0 animations:^{
    // Mess with bounds
} completion:^(BOOL f) {
    [imageView setImage:originalImage cacheName:@"cacheName"];
    [imageView setBoundingEnabled:YES];
}];
```

## Author

Kevin Donnelly - @donnellyk

## License

KVNBoundedImageView is available under the MIT license. See the LICENSE file for more info.

### Sample Photos

All example photos taken from Flickr licensed under The Creative Commons. Thanks to [Thomas Saito](http://www.flickr.com/photos/thomas-saito), [Franki](http://www.flickr.com/photos/francki), [Ableman](http://www.flickr.com/photos/ableman).


CHANGELOG.md
# KVNBoundedImageView CHANGELOG

## 0.1.0

Initial release.


Source/KVNBoundedImageView.h
//
//  KVNFaceAlignImageView.h
//  ImageTest
//
//  Created by Kevin Donnelly on 1/11/14.
//
//

#import <UIKit/UIKit.h>

extern NSString * const BoundingBoxSchemeAll; // Bounding rectangle uses all face features detected. A rectangle that that contains all feature bounds is used.
extern NSString * const BoundingBoxSchemeLargest; // Bounding rectangle uses only the largest face feature detected. The bounds of that feature is used. If two bounds are equal the last feature detected is used.
extern NSString * const BoundingBoxSchemeSmallest; // Bounding rectangle uses only the smallest face feature detected. The bounds of that feature is used. If two bounds are equal the last feature detected is used. Why would you ever want to use this, you ask? I have no idea, it was just easy to implement so I threw it in. Everyone loves an underdog!

@interface KVNBoundedImageView : UIImageView <NSURLConnectionDataDelegate>

@property CGFloat boundingPadding; // Padding added to each features bounds. Helpful for a face detected near the top of the image, as forheads are often not included. Default is 10.0.
@property (nonatomic, strong) NSString *detectorAccuracy; // Used as value for `CIDetectorAccuracy` key in `CIDetector` options. Default is `CIDetectorAccuracyLow`
@property (nonatomic, strong) NSString *boundingBoxScheme; // How the bounding rectangle is calculated. Default is `BoundingBoxSchemeAll`
@property (nonatomic) BOOL boundingEnabled; // If `NO`, acts like normal `UIImageView`. Defaults to `YES`.
@property (nonatomic) BOOL animated; // If YES, cropped image is animated into the view. Defaults to YES.

@property (nonatomic, strong, readonly) UIImage *originalImage; // Image before bounding/resizing
@property (nonatomic, strong) UIImage *placeholderImage; // Image to be shown before loading/cropping is completed. If an image is already displayed in the UIImageView, nothing is displayed.
@property (nonatomic, strong) NSString *nibImageCacheName; // If loading from nib, you can set this in the User Defined Runtime Attributes to cache the cropped image for reuse. Will be ignored for programmically set images.

/**
 *  Forces the current image to be resized again.
 */
- (void)fitToFeatures;

/**
 *  Stops URL loading and detection
 */
- (void)halt;

/**
 *  Clears cache of cropped and remotely loaded images
 */
+ (void)clearCache;

/**
 *  Bounding rectangle used to resize image. Override for alternative boundaries besides face detection. If subclassing, rectangle returned origin should be upper-left, as in `UIView` coordinate system.
 *
 *  @param image `UIImage` to build bounding rectangle from
 *
 *  @return `CGRect` of the bounding rectangle
 *
 *  @warning This method will be run on a background thread.
 */
- (CGRect)boundingRectangleForImage:(UIImage *)image;

/**
 *  Array of features found for previously set images. If overriding, return the individual rectangles, if any, that were used to calculate the bounding rectangle. The `CGRect` stored are in their original coordinate system
 *
 *  @return `NSArray` of `NSValue` objects that represent `CGRect`
 */
- (NSArray *)foundfeatures;

/**
 *  Remotely sets image from URL. Default `NSURLRequestCachePolicy` is used. Absolute URL is used for  the cropped image cache.
 
 *
 *  @param url URL where image is hosted
 */
- (void)setImageFromURL:(NSURL *)url placeholder:(UIImage *)placeholder;

/**
 *  Detects features, crops, and set UIImageView image, then caches results using name as key. If there is a valid cached version of the cropped image, it is set immediately
 *
 *  @param image Image to be presented
 *  @param cacheName  Key for image cache
 */
- (void)setImage:(UIImage *)image cacheName:(NSString *)cacheName;

/**
 *  Kicks off detection by setting the UIImage and displays the placeholder while the detection is ongoing.
 *
 *  @param image  Image to be displayed
 *  @param cacheName  Key for image cache
 *  @param placeholder Placeholder image to be displayed during detection and cropping
 */
- (void)setImage:(UIImage *)image cacheName:(NSString *)cacheName placeholder:(UIImage *)placeholder;
@end


Example/BoundedImageExample/BoundedImageExample/AppDelegate.m
//
//  AppDelegate.m
//  BoundedImageExample
//
//  Created by Kevin Donnelly on 2/9/14.
//
//

#import "AppDelegate.h"

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    // Override point for customization after application launch.
    return YES;
}
							
- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


Example/BoundedImageExample/BoundedImageExample/ViewController.m
//
//  ViewController.m
//  BoundedImageExample
//
//  Created by Kevin Donnelly on 2/9/14.
//
//

#import "ViewController.h"
#import "KVNBoundedImageView.h"

@interface ViewController ()

@end

@implementation ViewController
@synthesize pageOneBoundedImageView;
@synthesize pageTwoBoundedImageView;
@synthesize pageThreeBoundedImageView;
@synthesize pageFourBoundedImageView;

@synthesize pageControl;
@synthesize toggleButton;
@synthesize imageScrollView;

- (void)viewDidLoad {
    [super viewDidLoad];
	
    [imageScrollView setContentSize:CGSizeMake(CGRectGetWidth(imageScrollView.bounds) * 4, CGRectGetHeight(imageScrollView.bounds))];
    [self createProgrammicallyCreated];
    [self createLargest];
    [self createRemotelyLoaded];
}

- (void)createProgrammicallyCreated {
    KVNBoundedImageView *imageView = [[KVNBoundedImageView alloc] initWithFrame:CGRectMake(340, 175, 280, 200)];
    [imageView setImage:[UIImage imageNamed:@"guitar.jpg"] cacheName:@"guitar"];
    
    [imageScrollView addSubview:imageView];
    self.pageTwoBoundedImageView = imageView;
}

- (void)createLargest {
    KVNBoundedImageView *imageView = [[KVNBoundedImageView alloc] initWithFrame:CGRectMake(660, 75, 280, 350)];
    [imageView setBoundingBoxScheme:BoundingBoxSchemeLargest];
    [imageView setImage:[UIImage imageNamed:@"mask.jpg"]];
    // Not caching this image, if you tap the disable and then enable again, you can see the speed difference.
    
    [imageScrollView addSubview:imageView];
    self.pageThreeBoundedImageView = imageView;
}

- (void)createRemotelyLoaded {
    KVNBoundedImageView *imageView = [[KVNBoundedImageView alloc] initWithFrame:CGRectMake(980, 175, 280, 100)];
    [imageView setBoundingBoxScheme:BoundingBoxSchemeLargest];
    NSURL *url = [NSURL URLWithString:@"http://farm6.staticflickr.com/5336/7144867897_bc1d3b5277_c.jpg"];
    [imageView setImageFromURL:url placeholder:[UIImage imageNamed:@"placeholder"]];
    
    [imageView setBackgroundColor:[UIColor redColor]];
    
    [imageScrollView addSubview:imageView];
    self.pageFourBoundedImageView = imageView;
}

- (IBAction)toggleTap:(id)sender {
    BOOL enable = !pageFourBoundedImageView.boundingEnabled;
    
    [pageOneBoundedImageView setBoundingEnabled:enable];
    [pageTwoBoundedImageView setBoundingEnabled:enable];
    [pageThreeBoundedImageView setBoundingEnabled:enable];
    [pageFourBoundedImageView setBoundingEnabled:enable];
    
    [toggleButton setTitle:(enable) ? @"Disable" : @"Enable" forState:UIControlStateNormal];
}

- (IBAction)reloadImage:(id)sender {
    [pageFourBoundedImageView removeFromSuperview];
    pageFourBoundedImageView = nil;
    
    [KVNBoundedImageView clearCache];
    [self createRemotelyLoaded];
}

#pragma mark - UIScrollViewDelegate methods
- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    self.pageControl.currentPage = lround(scrollView.contentOffset.x / CGRectGetWidth(scrollView.bounds));
}

@end


Source/KVNBoundedImageView.m
//
//  KVNFaceAlignImageView.m
//  ImageTest
//
//  Created by Kevin Donnelly on 1/11/14.
//
//

#import "KVNBoundedImageView.h"

NSString * const BoundingBoxSchemeAll = @"BoundingBoxSchemeAll";
NSString * const BoundingBoxSchemeLargest = @"BoundingBoxSchemeLargest";
NSString * const BoundingBoxSchemeSmallest = @"BoundingBoxSchemeSmallest";

static CGFloat const FeaturePaddingDefault = 5.0;

@interface KVNBoundedImageView()

@property (strong) NSArray *features;

@property (nonatomic, strong) NSOperationQueue *asyncQueue;

// Properties related to URL image loading
@property (nonatomic, strong) NSOperation *imageFetchOperation;
@property (strong) NSURLRequest *imageURLRequest;
@property (nonatomic, strong) NSMutableData *imageData;
@property (nonatomic, strong) NSPort *runLoopPort;

@property CGSize latestBounds;

@property (nonatomic, strong) NSString *lastImageCacheName;

@end

@implementation KVNBoundedImageView
@synthesize detectorAccuracy = _detectorAccuracy;
@synthesize boundingBoxScheme = _boundingBoxScheme;


#pragma mark - Class Methods
+ (NSCache *)sharedCroppedImageCache {
    static NSCache *_sharedCroppedImageCache = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedCroppedImageCache = [[NSCache alloc] init];
    });
    
    return _sharedCroppedImageCache;
}

#pragma mark - Lifecycle
- (id)init {
    if (!(self = [super init])) {
        return nil;
    }
    
    [self commonSetup];
    return self;
}

- (id)initWithFrame:(CGRect)frame {
    if (!(self = [super initWithFrame:frame])) {
        return nil;
    }
    
    [self commonSetup];
    return self;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    [self commonSetup];
    
    self.lastImageCacheName = self.nibImageCacheName;
    [self fitToFeatures];
}

- (void)dealloc {
    [self halt];
    self.asyncQueue = nil;
}

- (void)commonSetup {
    self.asyncQueue = [[NSOperationQueue alloc] init];
    [self.asyncQueue setMaxConcurrentOperationCount:1];
    
    _detectorAccuracy = CIDetectorAccuracyLow;
    _boundingBoxScheme = BoundingBoxSchemeAll;
    _boundingEnabled = YES;
    _boundingPadding = FeaturePaddingDefault;
    _animated = YES;
    
    _latestBounds = self.bounds.size;
    
    self.clipsToBounds = YES;
    [self setContentMode:UIViewContentModeScaleAspectFill];
}

#pragma mark - Layout
- (void)layoutSubviews {
    [super layoutSubviews];
    if (!self.boundingEnabled) {
        return;
    }
    
    if (AspectRatio(self.bounds.size) != AspectRatio(self.latestBounds)) {
        [self fitToFeatures];
    }
}

- (CGRect)boundingRectangleForImage:(UIImage *)image {
    CGRect boundingRectangle = CGRectNull;
    
    if (!self.features) {
        self.features = [self findFeaturesFromImage:image];
    }
    
    BOOL useAll = [self.boundingBoxScheme isEqualToString:BoundingBoxSchemeAll];
    for (NSValue *feature in self.features) {
        CGRect bounds = [feature CGRectValue];
        if (useAll) {
            bounds = CGRectInset(bounds, -self.boundingPadding, -self.boundingPadding);
            
            if (CGRectIsNull(boundingRectangle)) {
                boundingRectangle = bounds;
            } else {
                boundingRectangle = CGRectUnion(boundingRectangle, bounds);
            }
        } else {
            CGFloat boundArea = CGRectGetWidth(boundingRectangle) * CGRectGetHeight(boundingRectangle);
            CGFloat featureBoundArea = CGRectGetWidth(bounds) * CGRectGetHeight(bounds);
            
            if ([self.boundingBoxScheme isEqualToString:BoundingBoxSchemeLargest]) {
                if (featureBoundArea >= boundArea) {
                    boundingRectangle = CGRectInset(bounds, -self.boundingPadding, -self.boundingPadding);
                }
            } else {
                if (CGRectIsNull(boundingRectangle) || featureBoundArea <= boundArea) {
                    boundingRectangle = CGRectInset(bounds, -self.boundingPadding, -self.boundingPadding);
                }
            }
            
        }
    }
    
    return boundingRectangle;
}

#pragma mark - Class methods
+ (void)clearCache {
    [[KVNBoundedImageView sharedCroppedImageCache] removeAllObjects];
}

#pragma mark - Image Resizing
- (void)fitToFeatures {
    [self setImage:(self.originalImage) ? self.originalImage : self.image cacheName:self.lastImageCacheName];
}

- (void)halt {
    [self.asyncQueue cancelAllOperations];
}

- (void)resizeImage:(UIImage *)image name:(NSString *)name completion:(void(^)(UIImage *croppedImage))completion {
    if (!self.boundingEnabled) {
        if (completion) {
            completion(image);
        }
    }
    
    __weak __typeof__(self)weakSelf = self;
    __block NSBlockOperation *blockOp = [NSBlockOperation blockOperationWithBlock:^() {
        if (!weakSelf) {
            return;
        }
        
        __strong __typeof__(weakSelf)strongSelf = weakSelf;
        
        CGRect boundingRect = [strongSelf boundingRectangleForImage:image];
        
        if ([blockOp isCancelled]) {
            return;
        }
    
        if (CGRectIsEmpty(boundingRect)) {
            [[NSOperationQueue mainQueue] addOperationWithBlock:^() {
                if (completion) {
                    completion(image);
                }
            }];
        }
        
        CGRect croppingRect = CGRectZero;
        CGFloat aspectRatio = AspectRatio(self.bounds.size);
        if (aspectRatio > 1.0) { // Scale by width
            croppingRect.size.width = image.size.width;
            croppingRect.size.height = CGRectGetWidth(croppingRect) / aspectRatio;
        } else { // Scale by height
            croppingRect.size.height = image.size.height;
            croppingRect.size.width = CGRectGetHeight(croppingRect) * aspectRatio;
        }
        
        CGFloat originX = MIN(MAX(0, CGRectGetMidX(boundingRect) - CGRectGetMidX(croppingRect)), image.size.width - CGRectGetWidth(croppingRect));
        CGFloat originY = MIN(MAX(0, CGRectGetMidY(boundingRect) - CGRectGetMidY(croppingRect)), image.size.height - CGRectGetHeight(croppingRect));
        croppingRect.origin = CGPointMake(originX, originY);
        
        // Flipping to CGImage Coordinate System
        croppingRect.origin.y = image.size.height - croppingRect.origin.y - croppingRect.size.height;
        
        if ([blockOp isCancelled]) {
            return;
        }
    
        CGImageRef imageRef = CGImageCreateWithImageInRect(image.CGImage, croppingRect);
        UIImage *croppedImage = [[UIImage alloc] initWithCGImage:imageRef];
        CGImageRelease(imageRef);
        
        if (name) {
            [[KVNBoundedImageView sharedCroppedImageCache] setObject:croppedImage forKey:[self cacheKeyWithName:name]];
        }
        
        if ([blockOp isCancelled]) {
            return;
        }
    
        [[NSOperationQueue mainQueue] addOperationWithBlock:^() {
            if (completion) {
                completion(croppedImage);
            }
        }];
    }];

    [self.asyncQueue addOperation:blockOp];
}

- (void)setCroppedImage:(UIImage *)image {
    [super setImage:image];
    
    if (self.animated) {
        CATransition *transition = [CATransition animation];
        transition.duration = .3f;
        transition.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
        transition.type = kCATransitionFade;
        
        [self.layer addAnimation:transition forKey:nil];
    }
}

#pragma mark - Face Detection
- (NSArray *)findFeaturesFromImage:(UIImage *)image {
    NSMutableArray *valueFeatures = [NSMutableArray array];

    // Not reusing the detector because that seemed to cause unbounded memory growth. until the detector was released. This was a workaround for that. http://openradar.appspot.com/radar?id=6645353252126720
    CIDetector *detector = [CIDetector detectorOfType:CIDetectorTypeFace context:nil options:@{CIDetectorAccuracy: self.detectorAccuracy}];
    
    CIImage *ciImage = [[CIImage alloc] initWithImage:image];
    NSArray *array = [detector featuresInImage:ciImage];
    
    for (CIFaceFeature *feature in array) {
        [valueFeatures addObject:[NSValue valueWithCGRect:feature.bounds]];
    }
    
    return valueFeatures;
}

- (NSArray *)foundfeatures {
    return self.features;
}

#pragma mark - Synthesis Override
- (void)setImage:(UIImage *)image {
    [self setImage:image cacheName:nil];
}

- (void)setPlaceholderImage:(UIImage *)placeholderImage {
    _placeholderImage = placeholderImage;
    
    if (!self.image) {
        [super setImage:placeholderImage];
    }
}

- (void)setBoundingBoxScheme:(NSString *)boundingBoxScheme {
    _boundingBoxScheme = boundingBoxScheme;
    [[KVNBoundedImageView sharedCroppedImageCache] removeObjectForKey:[self cacheKeyWithName:self.lastImageCacheName]];
    
    if (self.originalImage) {
        [self setImage:self.originalImage cacheName:self.lastImageCacheName];
    }
}

- (void)setDetectorAccuracy:(NSString *)detectorAccuracy {
    if (_detectorAccuracy == detectorAccuracy) {
        return;
    }
    [[KVNBoundedImageView sharedCroppedImageCache] removeObjectForKey:[self cacheKeyWithName:self.lastImageCacheName]];
    
    _detectorAccuracy = detectorAccuracy;
    
    if (self.originalImage) {
        [self setImage:self.originalImage cacheName:self.lastImageCacheName];
    }
}

- (void)setBoundingEnabled:(BOOL)boundingEnabled {
    _boundingEnabled = boundingEnabled;
    
    if (self.originalImage) {
        [self setImage:self.originalImage cacheName:self.lastImageCacheName];
    }
}

- (void)setContentMode:(UIViewContentMode)contentMode {
    [super setContentMode:UIViewContentModeScaleAspectFill];
}

#pragma mark - Setting Image
- (void)setImageFromURL:(NSURL *)url placeholder:(UIImage *)placeholder {
    [self setPlaceholderImage:placeholder];
    
    [self.asyncQueue cancelAllOperations];
    self.imageData = [[NSMutableData alloc] init];
    
    UIImage *cachedImage = [[KVNBoundedImageView sharedCroppedImageCache] objectForKey:[self cacheKeyWithName:[url absoluteString]]];
    
    if (cachedImage) {
        [super setImage:cachedImage];
        [self  setLastImageCacheName:[url absoluteString]];
        return;
    }
    
    __weak __typeof__(self)weakSelf = self;
    self.imageFetchOperation = [NSBlockOperation blockOperationWithBlock:^(){
        __strong __typeof__(weakSelf)strongSelf = weakSelf;

        if (!strongSelf) {
            return;
        }
        
        self.imageURLRequest = [NSURLRequest requestWithURL:url];
        NSURLConnection *connection = [[NSURLConnection alloc] initWithRequest:self.imageURLRequest delegate:strongSelf startImmediately:NO];
        
        NSRunLoop *runLoop = [NSRunLoop currentRunLoop];
        strongSelf.runLoopPort = [NSPort port];
        [runLoop addPort:strongSelf.runLoopPort forMode:NSDefaultRunLoopMode];
        
        [connection scheduleInRunLoop:runLoop forMode:NSRunLoopCommonModes];
        [connection start];
        [runLoop run];
        
    }];
    
    [self.asyncQueue addOperation:self.imageFetchOperation];
}

- (void)setImage:(UIImage *)image cacheName:(NSString *)cacheName {
    _originalImage = image;
    [self halt];
    self.imageURLRequest = nil;
    
    UIImage *cachedCroppedImage;
    if (!self.boundingEnabled) {
        [super setImage:image];
        return;
    } else if ((cachedCroppedImage = [[KVNBoundedImageView sharedCroppedImageCache] objectForKey:[self cacheKeyWithName:cacheName]])) {
        [super setImage:cachedCroppedImage];
        return;
    } else if (!self.image && self.placeholderImage) {
        [super setImage:self.placeholderImage];
    }
    
    __weak __typeof__(self)weakSelf = self;
    [self resizeImage:image name:(NSString *)cacheName completion:^(UIImage *croppedImage) {
        __strong __typeof__(weakSelf)strongSelf = weakSelf;
        [strongSelf setCroppedImage:croppedImage];
        strongSelf.lastImageCacheName = cacheName;
    }];
}

- (void)setImage:(UIImage *)image cacheName:(NSString *)cacheName placeholder:(UIImage *)placeholder {
    [self setPlaceholderImage:placeholder];
    [self setImage:image cacheName:cacheName];
}

#pragma mark NSURLConnectionDataDelegate
- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
    if ([self.imageFetchOperation isCancelled]) {
        [connection cancel];
        [self stopRunLoop];
        return;
    }
    
    [self.imageData appendData:data];
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
    [self stopRunLoop];
    
    UIImage *image = [UIImage imageWithData:self.imageData];
    if ([self.imageFetchOperation isCancelled]) {
        return;
    }
    
    __block NSString *cacheName = [self.imageURLRequest.URL absoluteString];
    [[NSOperationQueue mainQueue] addOperationWithBlock:^() {
        [self setImage:image cacheName:cacheName];
    }];
    
    self.imageData = nil;
    self.imageURLRequest = nil;
}

- (void)stopRunLoop {
    [[NSRunLoop currentRunLoop] removePort:self.runLoopPort forMode:NSDefaultRunLoopMode];
    self.runLoopPort = nil;
}

#pragma mark - Private Convenience Methods
- (NSString *)cacheKeyWithName:(NSString *)name {
    if (!name || !self.detectorAccuracy || !self.boundingBoxScheme) {
        return nil;
    }
    
    NSNumber *aspectRatio = [NSNumber numberWithFloat:AspectRatio(self.bounds.size)];
    return [NSString stringWithFormat:@"%@+%@+%@+%@", name, [aspectRatio stringValue], self.detectorAccuracy, self.boundingBoxScheme];
}

static inline CGFloat AspectRatio(CGSize size) {
    return size.width / size.height;
}

@end